File: python/bubble_sort.py
# Bubble Sort
def BubbleSort(arr : list[int]) -> list[int]:
	swaps = True
	while swaps:
		swaps = False
		i = 1
		while(i < len(arr)):
			if arr[i - 1] > arr[i]:
				tmp = arr[i]
				arr[i] = arr[i - 1]
				arr[i - 1] = tmp
				swaps = True
			i += 1
	return arr
	
"""
	arr = [12, 3, 2, 5, 10]
	
	step 1: [3, 12, 2, 5, 10]
	
	step 2: [3, 2, 12, 5, 10]
	
	step 3: [3, 2, 5, 12, 10]
	
	step 4: [3, 2, 5, 10, 12]


---------------------------------

	Complexity(Worst case): 
		
		Time : O(N^2)
		Memory : O(1)
		
		Swaps : N^2

"""

File: c/bubble_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

const int size = 100;

int arr[size];

void RandomIntGenerator();
void BubbleSort(int* arr, int size);

int main(){
	srand(time(NULL));
	RandomIntGenerator();
	
	BubbleSort(arr, size);
	for(int i = 0; i < size; i++){
		printf("%d ", arr[i]);
	}
	printf("\n");
	return 0;
}

void RandomIntGenerator(){
	for(int i = 0; i < size; i++){
		arr[i] = rand() % size * 10;
	}
}


void BubbleSort(int* arr, int size){
	int swaps = 1; // True
	while(swaps){
		int i = 1;
		swaps = 0;
		while(i < size){
			if(arr[i - 1] > arr[i]){
				int tmp = arr[i];
				arr[i] = arr[i - 1];
				arr[i - 1] = tmp;
				
				swaps = 1;
			}
		i++;
		}
	}
}